hack/vendor.sh

```diff
@@ -1,14 +1,41 @@
 #!/usr/bin/env bash
-
-# This file is just wrapper around 'go mod vendor' tool.
+#
+# This file is just a wrapper around the 'go mod vendor' tool.
 # For updating dependencies you should change `vendor.mod` file in root of the
 # project.
 
 set -e
-set -x
 
 SCRIPTDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 "${SCRIPTDIR}"/go-mod-prepare.sh
 
-GO111MODULE=auto go mod tidy -modfile 'vendor.mod' -compat 1.18
-GO111MODULE=auto go mod vendor -modfile vendor.mod
+export GO111MODULE=on
+
+tidy() {
+	(
+		set -x
+		go mod tidy -modfile vendor.mod -compat 1.18
+	)
+}
```

Comment:
Is this the same as using `()` for the outer brackets?

```suggestion
tidy() (
	set -x
	go mod tidy -modfile vendor.mod -compat 1.18
)
```


Also wondering if we still need the `-compat` and/or if we can make that use the Go version in either `vendor.mod` or `go.mod` instead (so that we're sure they're in sync)

Reply:
I'm using subshells here to print the command we run, but not generate more noise when we hit the second function in the file. We could also write `set +x`, but I think the subshell is a bit more idiomatic.

Your invocation is a syntax error, a function always needs the curly brackets to delimit the body.

Reply:
According to [the docs](https://go.dev/ref/mod) (ctrl-f for `-compat`), not specifying `compat` is the same as the `go` directive in the `go.mod` minus one.

I have a couple thoughts on this. I haven't yet addressed the duplication as I was thinking about adding a single canonical location to track the minimum supported Go version and default Go version.

We could also have a formal policy of "we support the latest Go version minus one", which would be instituted by dropping `-compat` here, `go` from `go.mod` and `vendor.mod`, and writing it down somewhere for contributors/packagers.

Reply:
> Your invocation is a syntax error, a function always needs the curly brackets to delimit the body.

[Incorrect](https://cuddly-octo-palm-tree.com/posts/2021-10-31-better-bash-functions/); using `()` for the outer brackets is well-formed shell syntax. A [shell function](https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_09_05) body is _any_ [compound command.](https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_09_04)

Not only is
```sh
foo() ( ls )
```
valid, so is
```sh
bar() while false; do ls; done
```

Reply:
The latter form I'm familiar with, but TIL on the former:

```
foo=0
bar=0

foo() (
	foo=1
)

bar() {
	bar=1
}

foo
echo $foo
bar
echo $bar
```

This works even in the Almquist shell -- it looks like I have some shell scripts in many projects to tidy up :laughing: